File: src/test/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/statistikk/domene/bransjeprogram/BransjeTest.java
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.domene.bransjeprogram;

import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.TestData.*;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domene.Næringskode5Siffer;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domene.Orgnr;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domene.UnderenhetLegacy;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domene.bransjeprogram.ArbeidsmiljøportalenBransje;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domene.bransjeprogram.Bransje;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domene.bransjeprogram.Bransjeprogram;
import org.junit.jupiter.api.Test;

public class BransjeTest {
  @Test
  public void
      virksomhetTilhørerBransjeprogram__skal_gi_true_hvis_næringskode_starter_med_de_definerte_sifrene() {
    Bransje bransje = new Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "test", "12", "45");
    assertThat(bransje.inkludererVirksomhet(underenhetMedNæringskode("45512"))).isTrue();
  }

  @Test
  public void
      virksomhetTilhørerBransjeprogram__skal_gi_false_hvis_næringskode_ikke_starter_med_de_definerte_sifrene() {
    Bransje bransje = new Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "test", "12", "45");
    assertThat(bransje.inkludererVirksomhet(underenhetMedNæringskode("46512"))).isFalse();
  }

  @Test
  public void finnBransje__mapperBarnehageTilBarnehagebransjen() {
    Næringskode5Siffer enBarnehage = new Næringskode5Siffer("88911", "");
    Bransje barnehageBransjen = Bransjeprogram.finnBransje(enBarnehage).orElseThrow();

    assertThat(barnehageBransjen.getNavn()).isEqualTo("Barnehager");
  }

  @Test
  public void finnBransje__mapperBoligbyggelagTilByggebransjen() {
    Næringskode5Siffer etBoligbyggerlag = new Næringskode5Siffer("41101", "");
    Bransje byggebranasjen = Bransjeprogram.finnBransje(etBoligbyggerlag).orElseThrow();

    assertThat(byggebranasjen.getNavn()).isEqualTo("Bygg");
  }

  @Test
  public void finnBransje__mapperBroOgTunellbyggerTilAnleggsbransjen() {
    Næringskode5Siffer enByggerAvBroOgTunnel = new Næringskode5Siffer("42130", "");

    Bransje byggebransjen = Bransjeprogram.finnBransje(enByggerAvBroOgTunnel).orElseThrow();
    assertThat(byggebransjen.getNavn()).isEqualTo("Anlegg");
  }

  @Test
  public void finnBransje__mapperProdusentAvIskremTilNæringsmiddelindustrien() {
    Næringskode5Siffer enProdusentAvIskrem = new Næringskode5Siffer("10520", "");

    Bransje næringsmiddelindustrien = Bransjeprogram.finnBransje(enProdusentAvIskrem).orElseThrow();
    assertThat(næringsmiddelindustrien.getNavn()).isEqualTo("Næringsmiddelsindustrien");
  }

  @Test
  public void finnBransje__mapperSomatiskeSpesialsykehusTilSykehusbransjen() {
    Næringskode5Siffer etSomatiskSpesialsykehus = new Næringskode5Siffer("86102", "");

    Bransje sykehusbransjen = Bransjeprogram.finnBransje(etSomatiskSpesialsykehus).orElseThrow();
    assertThat(sykehusbransjen.getNavn()).isEqualTo("Sykehus");
  }

  @Test
  public void finnBransje__mapperSykehjemTilSykehjemsbransjen() {
    Næringskode5Siffer etSomatiskSykehjem = new Næringskode5Siffer("87102", "");

    Bransje sykehjemsbransjen = Bransjeprogram.finnBransje(etSomatiskSykehjem).orElseThrow();
    assertThat(sykehjemsbransjen.getNavn()).isEqualTo("Sykehjem");
  }

  @Test
  public void finnBransje__mapperTurbiltransportTilTransportbransjen() {
    Næringskode5Siffer enturbiltransportør = new Næringskode5Siffer("49392", "");

    Bransje sykehjemsbransjen = Bransjeprogram.finnBransje(enturbiltransportør).orElseThrow();
    assertThat(sykehjemsbransjen.getNavn()).isEqualTo("Rutebuss og persontrafikk (transport)");
  }

  @Test
  public void bransje__skal_ikke_godta_koder_med_lengde_utenom_5_og_2() {
    assertThrows(
        IllegalArgumentException.class,
        () -> new Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "navn", "123"));
  }

  @Test
  public void bransje__skal_ikke_godta_koder_med_forskjellige_lengder() {
    assertThrows(
        IllegalArgumentException.class,
        () -> new Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "navn", "12", "12345"));
  }

  @Test
  public void bransje__skal_godta_koder_med_lengde_2_eller_5() {
    new Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "navn", "12", "34", "56");
    new Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "navn", "12345", "34567", "56789");
  }

  @Test
  public void lengdePåNæringskoder__skal_returnere_riktig_lengde() {
    assertTrue(
        new Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "navn", "12345", "34567", "56789")
            .erDefinertPåFemsiffernivå());
    assertTrue(
        new Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "navn", "12", "34", "45")
            .erDefinertPåTosiffernivå());
  }

  @Test
  public void inkludererVirksomhet__skal_returnere_hvorvidt_virksomhetens_næring_er_i_bransjen() {
    Bransje bransje =
        new Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "navn", "12345", "34567", "56789");
    assertThat(bransje.inkludererVirksomhet(underenhetMedNæringskode("12345"))).isTrue();
    assertThat(bransje.inkludererVirksomhet(underenhetMedNæringskode("54321"))).isFalse();
  }

  private UnderenhetLegacy underenhetMedNæringskode(String næringskode) {
    return new UnderenhetLegacy(
        etOrgnr(), new Orgnr("053497180"), "Underenhet AS", enNæringskode5Siffer(næringskode), 40);
  }
}

File: src/test/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/statistikk/domene/bransjeprogram/BransjeTest.kt
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.domene.bransjeprogram

import no.nav.arbeidsgiver.sykefravarsstatistikk.api.TestData.enNæringskode
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.TestData.etOrgnr
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.BedreNæringskode
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.Orgnr
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.UnderenhetLegacy
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.bransjeprogram.ArbeidsmiljøportalenBransje
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.bransjeprogram.Bransje
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.bransjeprogram.Bransjeprogram.finnBransje
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test

class BransjeTest {
    @Test
    fun virksomhetTilhørerBransjeprogram__skal_gi_true_hvis_næringskode_starter_med_de_definerte_sifrene() {
        val bransje = Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "test", listOf("12", "45"))
        Assertions.assertThat(bransje.inkludererVirksomhet(underenhetMedNæringskode("45512"))).isTrue()
    }

    @Test
    fun virksomhetTilhørerBransjeprogram__skal_gi_false_hvis_næringskode_ikke_starter_med_de_definerte_sifrene() {
        val bransje = Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "test", listOf("12", "45"))
        Assertions.assertThat(bransje.inkludererVirksomhet(underenhetMedNæringskode("46512"))).isFalse()
    }

    @Test
    fun finnBransje__mapperBarnehageTilBarnehagebransjen() {
        val enBarnehage = BedreNæringskode("88911")
        val (_, navn) = finnBransje(enBarnehage).orElseThrow()
        Assertions.assertThat(navn).isEqualTo("Barnehager")
    }

    @Test
    fun finnBransje__mapperBoligbyggelagTilByggebransjen() {
        val etBoligbyggerlag = BedreNæringskode("41101")
        val (_, navn) = finnBransje(etBoligbyggerlag).orElseThrow()
        Assertions.assertThat(navn).isEqualTo("Bygg")
    }

    @Test
    fun finnBransje__mapperBroOgTunellbyggerTilAnleggsbransjen() {
        val enByggerAvBroOgTunnel = BedreNæringskode("42130")
        val (_, navn) = finnBransje(enByggerAvBroOgTunnel).orElseThrow()
        Assertions.assertThat(navn).isEqualTo("Anlegg")
    }

    @Test
    fun finnBransje__mapperProdusentAvIskremTilNæringsmiddelindustrien() {
        val enProdusentAvIskrem = BedreNæringskode("10520")
        val (_, navn) = finnBransje(enProdusentAvIskrem).orElseThrow()
        Assertions.assertThat(navn).isEqualTo("Næringsmiddelsindustrien")
    }

    @Test
    fun finnBransje__mapperSomatiskeSpesialsykehusTilSykehusbransjen() {
        val etSomatiskSpesialsykehus = BedreNæringskode("86102")
        val (_, navn) = finnBransje(etSomatiskSpesialsykehus).orElseThrow()
        Assertions.assertThat(navn).isEqualTo("Sykehus")
    }

    @Test
    fun finnBransje__mapperSykehjemTilSykehjemsbransjen() {
        val etSomatiskSykehjem = BedreNæringskode("87102")
        val (_, navn) = finnBransje(etSomatiskSykehjem).orElseThrow()
        Assertions.assertThat(navn).isEqualTo("Sykehjem")
    }

    @Test
    fun finnBransje__mapperTurbiltransportTilTransportbransjen() {
        val enturbiltransportør = BedreNæringskode("49392")
        val (_, navn) = finnBransje(enturbiltransportør).orElseThrow()
        Assertions.assertThat(navn).isEqualTo("Rutebuss og persontrafikk (transport)")
    }

    @Test
    fun bransje__skal_ikke_godta_koder_med_lengde_utenom_5_og_2() {
        org.junit.jupiter.api.Assertions.assertThrows(
            IllegalArgumentException::class.java
        ) { Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "navn", listOf("123")) }
    }

    @Test
    fun bransje__skal_ikke_godta_koder_med_forskjellige_lengder() {
        org.junit.jupiter.api.Assertions.assertThrows(
            IllegalArgumentException::class.java
        ) { Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "navn", listOf("12", "12345")) }
    }

    @Test
    fun bransje__skal_godta_koder_med_lengde_2_eller_5() {
        Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "navn", listOf("12", "34", "56"))
        Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "navn", listOf("12345", "34567", "56789"))
    }

    @Test
    fun lengdePåNæringskoder__skal_returnere_riktig_lengde() {
        org.junit.jupiter.api.Assertions.assertTrue(
            Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "navn", listOf("12345", "34567", "56789"))
                .erDefinertPåFemsiffernivå()
        )
        org.junit.jupiter.api.Assertions.assertTrue(
            Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "navn", listOf("12", "34", "45"))
                .erDefinertPåTosiffernivå()
        )
    }

    @Test
    fun inkludererVirksomhet__skal_returnere_hvorvidt_virksomhetens_næring_er_i_bransjen() {
        val bransje = Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "navn", listOf("12345", "34567", "56789"))
        Assertions.assertThat(bransje.inkludererVirksomhet(underenhetMedNæringskode("12345"))).isTrue()
        Assertions.assertThat(bransje.inkludererVirksomhet(underenhetMedNæringskode("00000"))).isFalse()
    }

    private fun underenhetMedNæringskode(næringskode: String): UnderenhetLegacy {
        return UnderenhetLegacy(
            etOrgnr(), Orgnr("053497180"), "Underenhet AS", enNæringskode(næringskode), 40
        )
    }
}
